src/adera/Machines/Container.h

```diff
@@ -0,0 +1,139 @@
+
+    static void update_containers(osp::active::ActiveScene& rScene);
+
+    static MachineContainer& instantiate(
+            osp::active::ActiveScene& rScene,
+            osp::active::ActiveEnt ent,
+            osp::PCompMachine const& config,
+            osp::BlueprintMachine const& settings);
+
+}; // class SysMachineContainer
+
+//-----------------------------------------------------------------------------
+
+class MachineContainer : public osp::active::Machine
+{
+    friend SysMachineContainer;
+
+public:
+
+    static inline std::string smc_mach_name = "Container";
+
+    MachineContainer(osp::active::ActiveEnt ownID, float capacity, ShipResource resource);
+    MachineContainer(MachineContainer&& move) noexcept;
+    MachineContainer& operator=(MachineContainer&& move) noexcept;
+    ~MachineContainer() = default;
+
+    void propagate_output(osp::active::WireOutput *output) override;
+
+    osp::active::WireInput* request_input(osp::WireInPort port) override;
+    osp::active::WireOutput* request_output(osp::WireOutPort port) override;
+
+    std::vector<osp::active::WireInput*> existing_inputs() override;
+    std::vector<osp::active::WireOutput*> existing_outputs() override;
+
+    constexpr const ShipResource& check_contents() const
+    { return m_contents; }
+
+    /**
+     * Request a quantity of the contained resource
+     *
+     * Since the resources are stored as unsigned integers, avoiding wraparound
+     * is crucial. This function wraps the resource withdrawal process by
+     * internally checking the requested quantity of resource, bounds checking
+     * it, and only returning as much resources as are available.
+     *
+     * @param quantity [in] The requested quantity of resource
+     *
+     * @return The amount of resource that was received
+     */
+    uint64_t request_contents(uint64_t quantity);
+
+    /**
+     * Compute the current mass of the container contents
+     * @return The current mass, in kg
+     */
+    float compute_mass() const noexcept;
+private:
+    //osp::active::WireInput m_inputs;
+    osp::active::WireOutput m_outputs;
+
+    float m_capacity;
+    ShipResource m_contents;
+}; // class MachineContainer
+
+/* MachineContainer */
+inline MachineContainer::MachineContainer(osp::active::ActiveEnt ownID,
+    float capacity, ShipResource resource)
+    : Machine(true)
+    , m_capacity(capacity)
+    , m_contents(resource)
+    , m_outputs(this, "output")
+{
+    m_outputs.value() = osp::active::wiretype::Pipe{ownID};
+}
+
+inline MachineContainer::MachineContainer(MachineContainer&& move) noexcept
+    : Machine(std::move(move))
+    , m_capacity(std::move(move.m_capacity))
+    , m_contents(std::move(move.m_contents))
+    , m_outputs(this, std::move(move.m_outputs))
+{}
+
+inline MachineContainer& MachineContainer::operator=(MachineContainer&& move) noexcept
+{
+    m_enable = std::exchange(move.m_enable, false);
+    m_capacity = std::exchange(move.m_capacity, 0.0f);
```

Comment:
You'll want to use std::exchange in the copy-constructor as well. Uniformity is important :-)

Reply:
These weird move constructors are only needed because of the current wiring stuff. The new wiring PR removes all of these.